### PR TITLE
[CPU] Sporadic SIGFAULT in GridSample.

### PR DESCRIPTION
**Pull request by nshchego**
### Details:
 - *Fix for sporadic SIGFAULT in GridSample.*
 - *The relative threshold for bf16 was changed to 0.01*

### Tickets:
 - *94989*

### CODE REVIEW

**Comment by dmitry-gorokhov** [comment]
@yury-intel Could you please review the PR?

**Comment by nshchego** [comment]
@yury-intel, could you please take a look?

File: src/plugins/intel_cpu/src/nodes/kernels/jit_kernel_base.cpp [hunk 1]
```diff
@@ -286,26 +286,80 @@ void JitKernelBase::uni_vpbroadcastd(const Xbyak::Ymm &x, const Xbyak::Operand &
...
+            } else if (typeSize == 4) {
+                pinsrd(xmmDstMask, r32Ones, j);
+            } else if (typeSize == 8) {
+                pinsrq(xmmDstMask, r64Ones, j);
+            }
+        }
+        cmp(rWorkRest, elPerVec);
+        je(lEnd, T_NEAR);
+        L(lPerm);
+        vperm2f128(ymmDstMask, ymmDstMask, ymmDstMask, 0x1);
```
Comment (yury-intel) [inline src/plugins/intel_cpu/src/nodes/kernels/jit_kernel_base.cpp:360]: Is it CMPLTPS ?

File: src/plugins/intel_cpu/src/nodes/kernels/grid_sample.cpp [hunk 2]
```diff
@@ -857,14 +858,20 @@ void GridSampleKernel<isa>::borderPadding(const Vmm& vCoordDst, const Vmm& vCoor
...
         uni_vdivps(vAux, vCoordDst, vSrcDimMul2Sub1F);
-        uni_vroundps(vAux, vAux, 0x3);                          // Truncation
-        uni_vfnmadd231ps(vCoordDst, vAux, vSrcDimMul2Sub1F);    // abs(x) % D21
+        uni_vroundps(vAux, vAux, 0x3);                       // Truncation
+        uni_vfnmadd231ps(vCoordDst, vAux, vSrcDimMul2Sub1F); // abs(x) % D21
+
+        // WA to avoid floating point error. The result must not exceed the divisor.
+        vcmpps(kAux, vSrcDimMul2Sub1F, vCoordDst, CMPLEPS);
+        uni_vmovups(vCoordDst | kAux, vZeros);
+        vrangeps(vCoordDst, vCoordDst, vZeros, 0x1);
```
[the remark this answers is not in the document]
Comment (nshchego) [reply src/plugins/intel_cpu/src/nodes/kernels/grid_sample.cpp:874]: No, `CMPLTPS` is aligned with `cmpps` instruction and has a different meaning.

File: src/plugins/intel_cpu/src/nodes/kernels/grid_sample.cpp [hunk 3]
```diff
@@ -877,12 +884,16 @@ void GridSampleKernel<x64::avx512_core>::reflectionPadding(const Vmm& vCoordDst,
         uni_vdivps(vAux, vCoordDst, vSrcDimMul2F);
         uni_vroundps(vAux, vAux, 0x3);                   // Truncation
         uni_vfnmadd231ps(vCoordDst, vAux, vSrcDimMul2F); // (x % D2 + D2) % D2
+
+        // WA to avoid floating point error. The result must not exceed the divisor.
+        vcmpps(kAux, vSrcDimMul2F, vCoordDst, CMPLEPS);
+        uni_vmovups(vCoordDst | kAux, vZeros);
+        vrangeps(vCoordDst, vCoordDst, vZeros, 0x1);
```
[the remark this answers is not in the document]
Comment (nshchego) [reply src/plugins/intel_cpu/src/nodes/kernels/grid_sample.cpp:891]: The same as above.

File: src/plugins/intel_cpu/src/nodes/kernels/grid_sample.cpp [hunk 4]
```diff
@@ -6,6 +6,9 @@
 
 using namespace dnnl::impl::cpu;
 
+#define CMPLTPS 0x1
+#define CMPLEPS 0x2
```
[the remark this answers is not in the document]
Comment (nshchego) [reply src/plugins/intel_cpu/src/nodes/kernels/grid_sample.cpp:10]: These names are taken from [the table](https://www.felixcloutier.com/x86/cmpps#tbl-3-4) and look quite clear. For example: CMP - instruction name (Compare), LT - operation (Low Than), PS - data type.

File: src/plugins/intel_cpu/src/nodes/kernels/grid_sample.cpp [hunk 5]
```diff
@@ -925,6 +936,14 @@ void GridSampleKernel<isa>::reflectionPadding(const Vmm& vCoordDst, const Vmm& v
         uni_vdivps(vAux0, vCoordDst, vMul2Sub1);
         uni_vroundps(vAux0, vAux0, 0x3);               // Truncation
         uni_vfnmadd231ps(vCoordDst, vAux0, vMul2Sub1); // abs(x) % D21
+
+        // WA to avoid floating point error. The result must not exceed the divisor.
```
[the remark this answers is not in the document]
Comment (nshchego) [reply src/plugins/intel_cpu/src/nodes/kernels/grid_sample.cpp:940]: "Floating point error" - in general this error related to the type representation. Arithmetic operations for numbers with significantly different exponents cannot be done with desired accuracy. We need to be sure that "The result must not exceed the divisor" here and I don't see another solution for now.

File: src/plugins/intel_cpu/src/nodes/kernels/jit_kernel_base.cpp (hunk 1 above)
Comment (nshchego) [reply src/plugins/intel_cpu/src/nodes/kernels/jit_kernel_base.cpp:360]: This constant has another meaning.

File: src/plugins/intel_cpu/src/nodes/kernels/grid_sample.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (yury-intel) [reply src/plugins/intel_cpu/src/nodes/kernels/grid_sample.cpp:10]: Can you reflect it in comment?

File: src/plugins/intel_cpu/src/nodes/kernels/jit_kernel_base.cpp (hunk 1 above)
Comment (yury-intel) [reply src/plugins/intel_cpu/src/nodes/kernels/jit_kernel_base.cpp:360]: I suggest to reveal all magic numbers

File: src/plugins/intel_cpu/src/nodes/kernels/grid_sample.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (yury-intel) [reply src/plugins/intel_cpu/src/nodes/kernels/grid_sample.cpp:940]: If it's common solution for such kind of problem, we don't need in "WA". Looks ambiguously.

File: src/plugins/intel_cpu/src/nodes/kernels/grid_sample.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (nshchego) [reply src/plugins/intel_cpu/src/nodes/kernels/grid_sample.cpp]: Modified with comments.

File: src/plugins/intel_cpu/src/nodes/kernels/grid_sample.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (nshchego) [reply src/plugins/intel_cpu/src/nodes/kernels/grid_sample.cpp]: Modified

File: src/plugins/intel_cpu/src/nodes/kernels/jit_kernel_base.cpp (hunk 1 above)
Comment (nshchego) [reply src/plugins/intel_cpu/src/nodes/kernels/jit_kernel_base.cpp:360]: These immediate values do not have special names in documentation and I don't think that we can cover all of them for all instructions.

**Review by yury-intel — APPROVED** [review]

**Comment by nshchego** [comment]
@dmitry-gorokhov , please take a look.

**Review by dmitry-gorokhov — APPROVED** [review]

**Comment by ilya-lavrenov** [comment]
is it valid to port this change to 2022.3?

**Pull request merged**